# Linear System of Equations
---
What we focus on

- Inconsistency of linear systems
- Column view of linear systems
- Row echelon form and reduced row echelon form
- Gaussian elimination for finding REF
- Gauss-Jordan elimination for finding RREF
- Finding LU Decomposition of a matrix.
- Fininding inverse by Gauss-Jordan elimination

<p class="text-danger">Please contact the instructor if you encounter any errors on this page.</p>

--- 

### Linear systems, column view, and elementary transformations

Consider the following system of $m$ linear equations in $n$ variables.
$$\large{
\begin{aligned}
a_{11} x_1 + a_{12} x_2  + \cdots + a_{1n} x_n  &= b_1 \\
a_{21} x_1 + a_{22} x_2  + \cdots + a_{2n} x_n  &= b_2 \\
 \vdots \qquad \qquad   & \ \\
a_{m1} x_1 + a_{m2} x_2  + \cdots + a_{mn} x_n  &= b_ m
\end{aligned}}
$$
<br>
<div class="alert alert-block alert-success">
The solution of a linear system represents the <strong>point of intersection of hyperplanes</strong> given by the linear equations.
</div>

<p class="bg-info">
    A system of equations is called <strong>consistent</strong> if there exists at least one solution.
</p>
<img src="https://live.staticflickr.com/65535/53977675232_ee324c6b1e_z.jpg" />

<p class="bg-info">
    A system of equations is called <strong>inconsistent</strong> if there is no solution that is satisfied by all the equations.
    </p>

<img src="https://live.staticflickr.com/65535/53978997120_466dda2243_c.jpg" />


<div class="alert alert-block alert-success">
<strong> Column View of Linear Systems</strong>: The solution also represents the linear coding of the columns of a matrix $A$ to get a vector $\mathbf{b}$ in the column space ($\mathcal{C}(A)$).
 
$$
 x_1 \begin{pmatrix}a_{11}\\a_{21}\\ \vdots \\a_{m1}\end{pmatrix} +
 x_2 \begin{pmatrix}a_{12}\\a_{22}\\ \vdots \\a_{m2}\end{pmatrix} +
 \cdots +
 x_n \begin{pmatrix}a_{1n}\\a_{2n}\\ \vdots \\a_{mn}\end{pmatrix}
 =
 \begin{pmatrix}b_1\\b_2\\ \vdots \\b_m\end{pmatrix}
 $$
 
 or simply as
 $$\Large \color{blue}{
 x_1 A_{:1}+x_2 A_{:2}+ \cdots +x_n A_{:n} = \mathbf{b}
 }
 $$
 </div>

- The system could be represented in a compact form as $A\mathbf{x} = \mathbf{b}$, where 
 $$
 A=
\begin{pmatrix}
a_{11} & a_{12} & \cdots & a_{1n} \\
a_{21} & a_{22} & \cdots & a_{2n} \\
\vdots & \vdots & \ddots & \vdots \\
a_{m1} & a_{m2} & \cdots & a_{mn}
\end{pmatrix},\quad
\mathbf{x}=
\begin{pmatrix}
x_1 \\
x_2 \\
\vdots \\
x_n
\end{pmatrix},\quad
\mathbf{b}=
\begin{pmatrix}
b_1 \\
b_2 \\
\vdots \\
b_m
\end{pmatrix}
$$

- The system is called...

  - **overdetermined** when $m > n$. Generally, such systems do not have solutions. One can instead, use a least square approach to find the best possible approximation to a solution that minimizes the sum of square of errors. We will study least square problems in a future chapter.
  
  - **underdetermined** when $ m < n$. Such systems, if consistent, have infinitely many solutions. One can then constrain the solution space by using various forms or penalty (regularization) on the size.

#### Elementary Row Operation

<div class="alert alert-block alert-info">
The following operation on the rows of a matrix are called elementary row operations.

<p class="bg-info">1. Swap two rows.</p>
    
<p class="bg-info">2. Multiply a row by a non-zero scalar.</p>
    
<p class="bg-info">3. Add or subtract a non-zero multiple of one row to another.</p>
    
</div>

It should be noted that the elementary row operations on an augmented matrix $\left[A \; | \;\mathbf{b}\right]$, created from a linear system of equations, is associated with equivalent manipulations of respective equations in the linear system. These operations  do not change the solutions of the system.

<br>

## Gaussian Elimination Example

Gaussian elimination could be used to convert a matrix into row echelon form that has the following characteristics.

<div class="alert alert-block alert-success">
    
A matrix is said to be in <strong>row echelon form (REF)</strong> if the following holds
    <br>
<p class="bg-warning">1. All the zero-rows of the matrix are in the bottom.</p>
    
<p class="bg-warning">2. The first nonzero entry (pivot element) in any row is always to the right of the first non-zero entry in the previous row (if any). The column having the pivot element is called pivot column.</p>
<br>
A matrix is said to be in <strong>reduced row echelon form (RREF)</strong> if in addition to the above properties, the pivots columns are also in the <strong>proper form</strong>, meaning that the pivot element is one and all the other elements are zero in the pivot column. 
    
</div>

<br>

**Example:** Solve the following system of linear equations
$$
\begin{aligned}
2x_1 + 1x_2 + 3x_3 &= 3 \\
3x_1 - x_2 + 2x_3 &= -3 \\
-4x_1 - 3 x_2 + 6 x_3 &= 4
\end{aligned}
$$

We want to solve for $x_1$, $x_2$, and $x_3$ using the Gaussian elimination method.

Step 1: **Elimination**

Start with the <b> augmented matrix</b> $\left[A \; | \;\mathbf{b}\right]$, where $A$ is the coefficient matrix, and $\mathbf{b}$ is the vector of numbers on the right:

$$
\begin{matrix}
R_1\\R_2\\R_3
\end{matrix}
\begin{bmatrix}
2 & 1 & 3 & \vdots & 3 \\
3 & -1 & 2 & \vdots & -3 \\
-4 & -3 & 6 & \vdots & 4
\end{bmatrix}
$$

Perform Gaussian elimination to transform the matrix into <b>upper triangular form</b>:

$$
\begin{bmatrix}
2 & 1 & 3 & \vdots & 3 \\
0 & -5/2 & -5/2 & \vdots & -15/2 \\
0 & 0 & 13 & \vdots & 13
\end{bmatrix}
$$

First, eliminate to reduce to zeros the elements below the first pivot (in the first column):

$$
\begin{matrix}
R_1\\R_2\\R_3
\end{matrix}
\begin{bmatrix}
2 & 1 & 3 & \vdots & 3 \\
3 & -1 & 2 & \vdots & -3 \\
-4 & -3 & 6 & \vdots & 4
\end{bmatrix}
\xrightarrow[R_3 - \color{red}{(-2)} R_1]{R_2 - \color{red}{(\frac{3}{2})} R_1}
\begin{bmatrix}
2 & 1 & 3 & \vdots & 3 \\
0 & -5/2 & -5/2 & \vdots & -15/2 \\
0 & -1 & 12 & \vdots & 10
\end{bmatrix}
$$

Next, eliminate  to reduce to zeros the elements below the second pivot (in the second column):

$$
\xrightarrow[]{R_3 - \color{red}{(\frac{2}{5})} R_2}
\begin{bmatrix}
2 & 1 & 3 & | & 3 \\
0 & -5/2 & -5/2 & | & -15/2 \\
0 & 0 & 13 & | & 13
\end{bmatrix}
$$

Note 1: It should be noted that the above matrix is the augmented matrix for the system
$$
\begin{aligned}
2x_1 + 1x_2 + 3x_3 &= 3 \\
 -\frac{5}{2}x_2 - \frac{5}{2} x_3 &= -\frac{15}{2} \\
 13 x_3 &= 13
\end{aligned}
$$

Step 2: <b>Back Substitution</b> could be used to solve a system in upper triangular form.

Now, perform back substitution starting from the last equation:



$$13 x_3  = 13 \implies x_3 = 1$$

The the second equations gives
$$
-\frac{5}{2} x_2 - \frac{5}{2} x_3 = - \frac{15}{2} \implies -\frac{5}{2} x_2 = \left( - \frac{15}{2}+\frac{5}{2} x_3 \right) \implies x_2 = - \frac{2}{5} \left(-\frac{15}{2}+\frac{5}{1} (1) \right) =2
$$

And finally, the third equation is solved.
$$2 x_1 + x_2 + 3 x_3 = 3 \implies 2 x_1  = \left(3 - x_2 -3 x_3 \right) \implies x_1 = \frac{1}{2} \left(3-2-3(1) \right)= -1.
$$



Hence, the solution to the system is: $x_1 =-1$, $x_2 = 2$, and $x_3=1$.

<div class="alert alert-block alert-danger"> 
    
<strong>Note:</strong> Row-swapping would be required if the element in the pivot position (generally, the diagonal element in the pivot row) is zero.
    
</div>

Note: Take note of the multipliers above for use in the LU factorization as $\color{red}{l_{21}=\frac{3}{2}}$, $\color{red}{l_{31}=-2}$, and $\color{red}{l_{32}=\frac{2}{5}}$.



### Back-substitution
---
One can solve a square linear system with upper triangular structure (REF) $${\LARGE U{\bf x} = {\bf b} }$$ by using back-substitution, where 
$$
U = \begin{bmatrix}
  u_{11} & u_{12} & u_{13} & \ldots &   u_{1n} \\
        0  & u_{22} & u_{23} & \ldots &   u_{2n} \\
        \vdots  &  \vdots       &  \ddots & \ddots &    \vdots \\
        0  &    0     &     \cdots    & u_{n-1,n-1} & u_{n-1,n} \\
        0 &     0    &   \cdots      &   \cdots     &   u_{n,n}
\end{bmatrix}
$$
The backward substitution algorithm finds the unknowns starting from the last $x_n$ to the first $x_1$.
$$
\begin{aligned}
  x_n &= \frac{b_n}{u_{n,n}}, \\
  x_{n-1} &= \frac{1}{u_{n-1,n-1}}\,  \left( b_2 - u_{n-1,n}  x_n \right), \\
      & \ \ \vdots \\
  x_k &= \frac{1}{u_{k,k}}\, \left(b_k - \sum\limits_{i=n, step = -1}^{k+1} u_{k,i} x_i \right).
\end{aligned}
$$

### LU Decomposition
---
If Gaussian elimination does not require row-swaps, one can find the LU decomposition of a matrix $A$ as
$$
\begin{align}
\underset{n\times n}{\LARGE \mathbf{A}} &= \underset{n\times n}{\LARGE \mathbf{L}} \quad \quad \underset{n\times n}{\LARGE \mathbf{U}} \\[2ex]
&= \begin{bmatrix}
l_{11} & 0 & 0 & \ldots & 0\\
l_{21}& l_{22} & 0 & \ldots & 0\\
l_{31}& l_{32} & l_{33} & \ldots & 0\\
\vdots & \vdots &\vdots & \ddots & \vdots \\
l_{n1} & l_{n2} & l_{n3} & \ldots & l_{nn}
\end{bmatrix}\begin{bmatrix}
u_{11} & u_{12} & u_{13} & \ldots & u_{1n}\\
0 & u_{22} & u_{23} & \ldots & u_{2n}\\
0 & 0 & u_{33} & \ldots & u_{3n}\\
\vdots & \vdots  &\vdots & \ddots & \vdots \\
0 & 0 & 0 & \ldots & u_{nn}
\end{bmatrix}
\end{align}
$$
where $L$ is a <b><font color="red">unit</font> lower triangular</b> matrix ($l_{ii}=1$ for all $i$) and $U$ is an <b>upper triangular</b> matrix.

$$\left[A \; | \;\mathbf{b}\right] = 
\begin{bmatrix}
2 & 1 & 3 & \vdots & 3 \\
3 & -1 & 2 & \vdots & -3 \\
-4 & -3 & 6 & \vdots & 4
\end{bmatrix}
\xrightarrow[\text{REF}]{\text{Gaussian elimination}}
\begin{bmatrix}
2 & 1 & 3 & \vdots & 3 \\
0 & -5/2 & -5/2 & \vdots & -15/2 \\
0 & 0 & 13 & \vdots & 13
\end{bmatrix}
$$

Then the LU factorization $A$ is given by 

$$
\begin{bmatrix}
2 & 1 & 3  \\
3 & -1 & 2 \\
-4 & -3 & 6 
\end{bmatrix} = L\;U =\begin{bmatrix}
1 & 0 & 0  \\
\color{red}{\frac{3}{2}} & 1 & 0  \\
\color{red}{-2} & \color{red}{\frac{2}{5}} & 1 
\end{bmatrix}
\begin{bmatrix}
2 & 1 & 3  \\
0 & -5/2 & -5/2  \\
0 & 0 & 13 
\end{bmatrix}
$$
<p bg="bg-info">
Note that the matrix $L$ is being created by using the multipliers in the process of Gaussian elimination in the example above $\color{red}{l_{21}=\frac{3}{2}}$, $\color{red}{l_{31}=-2}$, and $\color{red}{l_{32}=\frac{2}{5}}$.</p>

In addition, please note the following
    
$$\begin{bmatrix}
1 & 0 & 0  \\
0 & 1 & 0  \\
0 & - \frac{2}{5} & 1 \end{bmatrix}
\begin{bmatrix}
1 & 0 & 0  \\
0 & 1 & 0  \\
2 & 0 & 1 \end{bmatrix}
\begin{bmatrix}
1 & 0 & 0  \\
- \frac{3}{2} & 1 & 0  \\
0 & 0 & 1 \end{bmatrix}
\begin{bmatrix}
2 & 1 & 3  \\
3 & -1 & 2 \\
-4 & -3 & 6 \end{bmatrix}
=\begin{bmatrix}
2 & 1 & 3  \\
0 & -5/2 & -5/2  \\
0 & 0 & 13 \end{bmatrix}
$$
    
One can use the Gaussian elimination with <b>partial pivoting</b> (requires row-swaps) to decompose the matrix $A$ as follows
    
$$
P A = L U,\text{ or }\ A = P^T L U;
$$
    
where $P$ is a permutation matrix, $L$ is a unit lower-triangular matrix and $U$ is an upper triangular matrix.
    
<div class="alert alert-block alert-success">
    
  <strong>Application of LU Decomposition</strong> Once the LU decomposition of the coefficient matrix for a linear systems $A\mathbf{x}=\mathbf{b}$ is known, the quivalent system $LU \mathbf{x}=\mathbf{b}$ could be solved for any right hand side in $\mathcal{O}(n^2)$ operations.

        
Step 1: Solve $ L \mathbf{z}=\mathbf{b}$ for $\mathbf{z}$ by forward substitution.
    
Step 2: Then solve $ U \mathbf{x}=\mathbf{z}$ for $\mathbf{x}$ by backward substitution.
</div>

## Gauss-Jordan Elimination 
---
The main idea in Gauss-Jordan elimination is to bring the augmented matrix from the system to reduced-row-echelon-form (RREF) by using elementary row operations.

 
 <b>Example</b>: Find the general solution of the following linear system.
 $$
 \begin{matrix}
 x+2y-z&=1\\
 2x+4y+z&=2\\
 3x+6y-z&=3
 \end{matrix}
 $$
 
On applying Gauss-Jordan elimination and reducing to RREF the augmented matrix undergoes the following transformation:
$$ \begin{bmatrix}
1 & 2 & -1 & | & 1 \\
2 & 4 & 1 & | & 2 \\
3 & 6 & -1 & | & 3
\end{bmatrix} 
\xrightarrow[RREF]{}
\begin{bmatrix}
1 & 2 & 0 & | & 1 \\
0 & 0 & 1 & | & 0 \\
0 & 0 & 0 & | & 0
\end{bmatrix}
$$

The above being a **rank deficient system** has infinitely many solutions that could be given by the general formula

$$\mathbf{x} = 
\begin{pmatrix}
1\\0\\0
\end{pmatrix}+ k_1 \begin{pmatrix}
2\\-1\\0
\end{pmatrix}
$$

See the supplementary note on rectangular systems for the details on how to find such general solutions.

<div class="alert alert-danger"> <strong>Note</strong>: If any row in the augmented matrix, upon Gaussian elimination, ends up in the form $[0 \ 0 \ \cdots \ 0 \ | \ c]$ where $c$ is a nonzero scalar, then the systems is inconsistent, has no solutions.</div>

<b> Example:</b> Solve the following system by using Gauss Jordan Elimination

$$
\begin{matrix}
2x+4y&=6\\
5x-2y+6z&=9\\
3x+y-2z&=2
\end{matrix}
$$

The augmented matrix of the given system

$$
\begin{bmatrix}
  2. & 4. & 0. & 6.\\
  5. & -2. & 6. & 9.\\
  3. & 1. & -2. & 2.\\
\end{bmatrix}
$$

#### Step 1:
Perform row operations to create zeros below the leading entries in the first column. The goal is to have a leading 1 in the first column.

Pass 1:

$$
\begin{bmatrix}
  1. & 2. & 0. & 3.\\
  0. & -12. & 6. & -6.\\
  0. & -5. & -2. & -7.\\
\end{bmatrix}
$$

#### Step 2:
Perform row operations to create zeros below the leading entries in the second column. The goal is to have leading 1's in the first two columns.
Pass 2:

$$
\begin{bmatrix}
  1. & 0. & 1. & 2.\\
  -0. & 1. & -0.5 & 0.5\\
  0. & 0. & -4.5 & -4.5\\
\end{bmatrix}
$$

#### Step 3:
Perform row operations to create zeros above and below the leading entry in the third column. The goal is to have a leading 1 in the third column.
Pass 3:

$$
\begin{bmatrix}
  1. & 0. & 0. & 1.\\
  0. & 1. & 0. & 1.\\
  0. & 0. & 1. & 1.\\
\end{bmatrix}
$$


#### Result:
$$
\begin{bmatrix}
  2. & 4. & 0. & 6.\\
  5. & -2. & 6. & 9.\\
  3. & 1. & -2. & 2.\\
\end{bmatrix}
\xrightarrow[R_3 -(3) R_1^*]{R_1^* \leftarrow (1/2) R_1;\ R_2 - (5) R_1^*}
\begin{bmatrix}
  1. & 2. & 0. & 3.\\
  0. & -12. & 6. & -6.\\
  0. & -5. & -2. & -7.\\
\end{bmatrix}
$$
<br>
$$
\xrightarrow[R_3 - \color{red}{(-5)} R_2^*]{R_2^* \leftarrow (-1/12) R_2;\ R_1 - (2) R_2^*}
\begin{bmatrix}
  1. & 0. & 1. & 2.\\
  0. & 1. & -0.5 & 0.5\\
  0. & 0. & -4.5 & -4.5\\
\end{bmatrix}
\xrightarrow[R_2 - (-0.5) R_3^*]{R_3^* \leftarrow (-1/4.5) R_3;\  R_1 - (1) R_3^*}
\begin{bmatrix}
  1. & 0. & 0. & 1.\\
  -0. & 1. & 0. & 1.\\
  -0. & -0. & 1. & 1.\\
\end{bmatrix}
$$

The matrix is now in reduced row echelon form, and the solution to the system can be easily read from it. The solution is $x = 1$, $y = 1$, and $z = 1$.

In [2]:
import numpy as np
# Following function returns a matrix using Latex notation from NumPy Notation
def bmatrix(A):
    """This function converts a 1-2 dimensional array into a LaTeX B-matrix
    INPUT:a: numpy array
    OUTPUT: LaTeX B-matrix code
    """
    if len(A.shape) > 2:
        raise ValueError('The input array must be a vector or a matrix.')
    
    rows = str(A).replace('[', '').replace(']', '').splitlines()
    ltx_str = [r'\begin{bmatrix}']
    ltx_str += ['  ' + ' & '.join(val.split()) + r'\\' for val in rows]
    ltx_str +=  [r'\end{bmatrix}']
    return '\n'.join(ltx_str)

In [3]:
# Gauss-Jordan Elimination (without row-swapping)
def findRREF(A):
    '''
    Input: a general rectangular matrix
    Output: the row reduced echelon form of the matrix A 
    by using Gauss-Jordan elimination
    '''
    augA = np.copy(A)
    m,n=augA.shape
    print("Augmented matrix: \n",bmatrix(augA))
    for k in range(m):
        # Step 0: Check that the matrix is not rank deficient. Row-swappign may be
        # required otherwise
        if np.abs(augA[k,k]) < 10**(-16):
            print("The given matrix is singular, or requires row swapping")
            break
        # Step 1: Convert the pivot element to 1
        augA[k,:] = augA[k,:] /augA[k,k]

        for i in range(m):
            if i==k:
                continue
            z = -augA[i,k] # multiplier ensures that pivot column comes in the proper form
            # Change the entire rows but the current
            augA[i,:] = augA[i,:] + z*augA[k,:] 
        print("Pass {}:\n".format(k+1))
        print(bmatrix(augA))
    return augA

In [28]:
#Example
A = np.array([[1,-2,1,-5],
             [2,1,7,5],
             [1,-1,2,-2]],dtype=float)
print(findRREF(A))

Augmented matrix: 
 \begin{bmatrix}
  1. & -2. & 1. & -5.\\
  2. & 1. & 7. & 5.\\
  1. & -1. & 2. & -2.\\
\end{bmatrix}
Pass 1:

\begin{bmatrix}
  1. & -2. & 1. & -5.\\
  0. & 5. & 5. & 15.\\
  0. & 1. & 1. & 3.\\
\end{bmatrix}
Pass 2:

\begin{bmatrix}
  1. & 0. & 3. & 1.\\
  0. & 1. & 1. & 3.\\
  0. & 0. & 0. & 0.\\
\end{bmatrix}
The given matrix is singular, or requires row swapping
[[1. 0. 3. 1.]
 [0. 1. 1. 3.]
 [0. 0. 0. 0.]]


Augmented matrix: 

$$ \begin{bmatrix}
  1. & -2. & 1. & -5.\\
  2. & 1. & 7. & 5.\\
  1. & -1. & 2. & -2.\\
\end{bmatrix}$$

Pass 1:

$$
\begin{bmatrix}
  1. & -2. & 1. & -5.\\
  0. & 5. & 5. & 15.\\
  0. & 1. & 1. & 3.\\
\end{bmatrix}$$

Pass 2:

$$
\begin{bmatrix}
  1. & 0. & 3. & 1.\\
  0. & 1. & 1. & 3.\\
  0. & 0. & 0. & 0.\\
\end{bmatrix}
$$

In [ ]:
# Modify the above code to find inverse
def invByRREF(A):
    m,n=A.shape
    if m!=n:
        exit("To find inverse, please provide a square matrix only.")
    # Appropriate augmented matrix. One line
    augA = # Your code here 
    modified_augA = findRREF(augA)
    # Take appropriate slice of modified_augA for inverse
    invA = # Your code here
    return invA

#### Example: Solve the following system
$$
\begin{bmatrix}
2 & 4 & 0 \\
5 & -2 & 6 \\
3 & 1 & -2 \\
\end{bmatrix}\ \begin{bmatrix}
x_1\\x_2\\x_3
\end{bmatrix}=\begin{bmatrix}
6\\9\\2
\end{bmatrix}
$$

In [4]:
np.set_printoptions(precision=4)
A = np.array([[2,4,0,6],
             [5,-2,6,9],
             [3,1,-2,2]],dtype=float)
augA_modified = findRREF(A)
# The last column is the solution
soln = augA_modified[:,3:]
print("\n The Solution is \n",soln)

Augmented matrix: 
 \begin{bmatrix}
  2. & 4. & 0. & 6.\\
  5. & -2. & 6. & 9.\\
  3. & 1. & -2. & 2.\\
\end{bmatrix}
Pass 1:

\begin{bmatrix}
  1. & 2. & 0. & 3.\\
  0. & -12. & 6. & -6.\\
  0. & -5. & -2. & -7.\\
\end{bmatrix}
Pass 2:

\begin{bmatrix}
  1. & 0. & 1. & 2.\\
  -0. & 1. & -0.5 & 0.5\\
  0. & 0. & -4.5 & -4.5\\
\end{bmatrix}
Pass 3:

\begin{bmatrix}
  1. & 0. & 0. & 1.\\
  -0. & 1. & 0. & 1.\\
  -0. & -0. & 1. & 1.\\
\end{bmatrix}

 The Solution is 
 [[1.]
 [1.]
 [1.]]


In [15]:
np.set_printoptions(precision=4)
A = np.array([[2,4,0, 1, 0, 0],
             [0,3,6,0,1,0],
             [0,0,1,0,0,1]],dtype=float)
print("\n The Inverse is \n",bmatrix(findRREF(A)[:,3:]))

Augmented matrix: 
 \begin{bmatrix}
  2. & 4. & 0. & 1. & 0. & 0.\\
  0. & 3. & 6. & 0. & 1. & 0.\\
  0. & 0. & 1. & 0. & 0. & 1.\\
\end{bmatrix}
Pass 1:

\begin{bmatrix}
  1. & 2. & 0. & 0.5 & 0. & 0.\\
  0. & 3. & 6. & 0. & 1. & 0.\\
  0. & 0. & 1. & 0. & 0. & 1.\\
\end{bmatrix}
Pass 2:

\begin{bmatrix}
  1. & 0. & -4. & 0.5 & -0.6667 & 0.\\
  0. & 1. & 2. & 0. & 0.3333 & 0.\\
  0. & 0. & 1. & 0. & 0. & 1.\\
\end{bmatrix}
Pass 3:

\begin{bmatrix}
  1. & 0. & 0. & 0.5 & -0.6667 & 4.\\
  0. & 1. & 0. & 0. & 0.3333 & -2.\\
  0. & 0. & 1. & 0. & 0. & 1.\\
\end{bmatrix}

 The Inverse is 
 \begin{bmatrix}
  0.5 & -0.6667 & 4.\\
  0. & 0.3333 & -2.\\
  0. & 0. & 1.\\
\end{bmatrix}


### Inverse of a matrix by Gauss-Jordan: an Example
$$
\begin{bmatrix}
  2. & 4. & 0. & 1. & 0. & 0.\\
  0. & 3. & 6. & 0. & 1. & 0.\\
  0. & 0. & 1. & 0. & 0. & 1.\\
\end{bmatrix}
$$

Pass 1:

$$
\begin{bmatrix}
  1. & 2. & 0. & 0.5 & 0. & 0.\\
  0. & 3. & 6. & 0. & 1. & 0.\\
  0. & 0. & 1. & 0. & 0. & 1.\\
\end{bmatrix}
$$

Pass 2:

$$
\begin{bmatrix}
  1. & 0. & -4. & 0.5 & -0.6667 & 0.\\
  0. & 1. & 2. & 0. & 0.3333 & 0.\\
  0. & 0. & 1. & 0. & 0. & 1.\\
\end{bmatrix}
$$

Pass 3:

$$
\begin{bmatrix}
  1. & 0. & 0. & 0.5 & -0.6667 & 4.\\
  0. & 1. & 0. & 0. & 0.3333 & -2.\\
  0. & 0. & 1. & 0. & 0. & 1.\\
\end{bmatrix}
$$

 The Inverse is 
 
 $$
 \begin{bmatrix}
  0.5 & -0.6667 & 4.\\
  0. & 0.3333 & -2.\\
  0. & 0. & 1.\\
\end{bmatrix}=
\begin{bmatrix}
  1/2 & -2/3 & 4.\\
  0. & 1/3 & -2.\\
  0. & 0. & 1.\\
\end{bmatrix}
$$

**Question**: Can you modify the code for Gauss-Jordan elimination to do the following?
1. Solve a square system $A{\bf x}={\bf b}$
1. Find the inverse of a matrix
1. Find the determinant of a matrix
1. Find the rank of a matrix

# PLEASE DISREGARD THE FOLLOWING
---

In [9]:
# Just some checks
invL = np.array([[1,0,0,],[-3/2,1,0],[13/5,-2/5,1]],dtype=float)

In [10]:
np.linalg.inv(invL)

array([[ 1.0000e+00, -4.2701e-17,  0.0000e+00],
       [ 1.5000e+00,  1.0000e+00,  0.0000e+00],
       [-2.0000e+00,  4.0000e-01,  1.0000e+00]])

In [29]:
# Creating a permutation matrix P from a permutation vector p.
I = np.eye(4)
p=[2,0,3,1]
P=I[p,:]
print(P)

[[0. 0. 1. 0.]
 [1. 0. 0. 0.]
 [0. 0. 0. 1.]
 [0. 1. 0. 0.]]


In [37]:
# Finding norms using Numpy
v=np.arange(0,10,1)
np.linalg.norm(v,ord=2)

16.881943016134134

In [5]:
import numpy as np
augA = np.array([[1,2,3],
                  [4,5,6],
                  [7,8,9]],dtype='float')
augA[[0, 1]] = augA[[1, 0]]
print(augA)
print(augA[[0, 1]])

[[4. 5. 6.]
 [1. 2. 3.]
 [7. 8. 9.]]
[[4. 5. 6.]
 [1. 2. 3.]]


In [8]:
augA = np.array([[0,1,1,1],
                  [1,0,0,0],
                  [1,0,0,0],
                  [1,0,0,0]],dtype='float')

In [9]:
print(np.linalg.eig(augA))

(array([ 1.73205081, -1.73205081,  0.        ,  0.        ]), array([[ 0.70710678,  0.70710678,  0.        ,  0.        ],
       [ 0.40824829, -0.40824829, -0.57735027, -0.57735027],
       [ 0.40824829, -0.40824829,  0.78867513, -0.21132487],
       [ 0.40824829, -0.40824829, -0.21132487,  0.78867513]]))
